File: addter.pl
#!/usr/bin/perl
open(INPUT,"$ARGV[0]");
open(OUTPUT,">$ARGV[1]");
@molecule=<INPUT>;
$k=substr($molecule[0],22,4);
$res=0;
$atom=0;
for($i=0;$i<@molecule;$i++)
{$res_num=substr($molecule[$i],22,4);
if($k==$res_num)
{$residue[$res][$atom]=$molecule[$i];
$atom++;}
if($k!=$res_num)
{$res++;
$atom=0;
$residue[$res][$atom]=$molecule[$i];
$atom++;}
$k=$res_num;}
for($i=0;$i<=$res;$i++)
{$link=0;
foreach $line (@{$residue[$i]})
{if(substr($line,12,3) eq " C ")
{foreach $line2 (@{$residue[$i+1]})
{if(substr($line2,12,3) eq " N ")
{$x1=substr($line,30,8);
$y1=substr($line,38,8);
$z1=substr($line,46,8);
$x2=substr($line2,30,8);
$y2=substr($line2,38,8);
$z2=substr($line2,46,8);
$dist=sqrt(($x1-$x2)**2+($y1-$y2)**2+($z1-$z2)**2);
if($dist<1.5)
{$link=1;}}}}
print OUTPUT ($line);}
if($link==0)
{print OUTPUT ("TER\n");}}

File: addter.py
#!/usr/bin/python
import sys,os
usage="""
Add 'TER' to each end of peptide
Usage $0 in.pdb out.pdb
"""
if len(sys.argv)<3:
    print usage
    sys.exit()
infile,outfile=sys.argv[1:3]

sys.stderr.write("!!!!! Warning addter.py is not completed by now, make sure 'N' atom is the first of each residue\n")

Nxyz,Cxyz=(None,None)
Newresiflag=True
oldindex=0
oldline=""

def distance(A,B):
    if None in (A,B):
        return 0
    dist=( (A[0]-B[0])**2 + (A[1]-B[1])**2 + (A[2]-B[2])**2 ) ** 0.5 
    return dist 

ofp=open(outfile,"w")
for line in open(infile,"r"):
    if line[0:6] not in ("ATOM  " , "HETATM"):
        ofp.write(line)
        continue
    if line[12:16].strip()=="N":
        Nxyz=(float(line[30:38]),float(line[38:46]),float(line[46:54]))
    if line[12:16].strip()=="C":
        Cxyz=(float(line[30:38]),float(line[38:46]),float(line[46:54]))
    index=int(line[22:26])
    if index!=oldindex and distance(Nxyz,Cxyz)>2.0:
        ofp.write("TER\n")
    oldindex=index
    ofp.write(line)
ofp.close()
